codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Paginators.java

```diff
@@ -0,0 +1,250 @@
+
+                    StringBuilder nilGuard = new StringBuilder();
+                    StringBuilder outputPath = new StringBuilder("result");
+
+                    List<MemberShape> outputMemberPath = paginationInfo.getOutputTokenPath();
+                    for (int i = 0; i < outputMemberPath.size(); i++) {
+                        MemberShape memberShape = outputMemberPath.get(i);
+                        outputPath.append(".");
+                        outputPath.append(symbolProvider.toMemberName(memberShape));
+
+                        // Don't need to check the last member here as we won't dereference it
+                        if (i != outputMemberPath.size() - 1) {
+                            if (i != 0) {
+                                nilGuard.append(" && ");
+                            }
+                            nilGuard.append(outputPath);
+                            nilGuard.append(" != nil");
+                        }
+                    }
+
+                    Runnable setToken = () -> {
+                        writer.write("p.nextToken = $L", outputPath);
+                    };
+                    if (nilGuard.length() > 0) {
+                        writer.openBlock("if $L {", "}", nilGuard.toString(), setToken::run);
+                    } else {
+                        setToken.run();
+                    }
+
+                    writer.openBlock("if p.options.StopOnDuplicateToken && "
+                            + "params.$L != nil && p.nextToken != nil && "
+                            + "*params.$L == *p.nextToken {", "}", inputMember, inputMember, () -> {
+                        writer.write("p.nextToken = nil");
+                    });
+
+                    writer.write("");
+                    writer.write("return result, nil");
+                });
+    }
+
+    private void writePaginatorOptions(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            Symbol operationSymbol,
+            MemberShape pageSizeMember,
+            Symbol optionsSymbol
+    ) {
+        writer.writeDocs(String.format("%s is the paginator options for %s", optionsSymbol.getName(),
+                operationSymbol.getName()));
+        writer.openBlock("type $T struct {", "}", optionsSymbol, () -> {
+            if (pageSizeMember != null) {
+                pageSizeMember.getMemberTrait(model, DocumentationTrait.class).ifPresent(documentationTrait -> {
+                    writer.writeDocs(documentationTrait.getValue());
+                });
+                writer.write("Limit $P", symbolProvider.toSymbol(pageSizeMember));
+                writer.write("");
+            }
+            writer.writeDocs("Set to true if pagination should stop if the service returns a pagination token that "
+                    + "matches the most recent token provided to the service.");
+            writer.write("StopOnDuplicateToken bool");
```

Comment:
We probably want to only generated the `StopOnDuplicateToken` on string `nextToken` shapes since those are the ones that are comparable.